File: sample/castro/examples/sieve.c
// Sieve of Eratosthenes — count primes < N.
// Stresses: tight integer loops, array indexing, store/load mix.

int prime[200000];

int sieve(int n) {
    for (int i = 0; i < n; i++) prime[i] = 1;
    prime[0] = 0;
    prime[1] = 0;
    // Only walk the marker loop while i*i fits in int (and < n).
    for (int i = 2; i * i < n; i++) {
        if (prime[i]) {
            for (int j = i * i; j < n; j += i) prime[j] = 0;
        }
    }
    int count = 0;
    for (int i = 2; i < n; i++) if (prime[i]) count++;
    return count;
}

int main() {
    int total = 0;
    for (int k = 0; k < 8; k++) total += sieve(200000);
    return total & 0xFF;
}

File: sample/jstro/benchmark/sieve.js
// Sieve of Eratosthenes — array-heavy benchmark.
function sieve(n) {
  var a = new Array(n + 1);
  for (var i = 0; i <= n; i++) a[i] = true;
  a[0] = false; a[1] = false;
  var count = 0;
  for (var i = 2; i <= n; i++) {
    if (a[i]) {
      count++;
      for (var j = i * 2; j <= n; j += i) a[j] = false;
    }
  }
  return count;
}

var n = 1000000;
var start = Date.now();
var c = sieve(n);
var elapsed = (Date.now() - start) / 1000;
console.log("primes <= " + n + " = " + c);
console.log("elapsed: " + elapsed + "s");
